File: live402/batch_profiles/algorand_manifest.py
"""Versioned Algorand payment manifests; jobs and payments are separate counts.

A quote is an observed merchant declaration, not independent chain verification.
The buyer must read current suggested parameters before signing. Per-byte fee
markets are deliberately refused by this initial bounded quote profile.
"""
from live402 import route_binding as rb
from live402.batch_profiles.algorand import NETWORK, ASSET, EXTENSION, _address, _uint
import re

ATOMIC = "algorand-atomic-multi-item-v1"
INVOICE = "algorand-aggregate-invoice-v1"
PROFILES = (ATOMIC, INVOICE)
COMMON_KEYS = {"network", "asset", "recipient", "fee_payer", "max_total_amount_atomic", "max_sponsor_fee_micro_algo", "job_hashes"}
GENESIS = "wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8="


def check(value):
    if not value:
        raise rb.BindingError("unsupported_algorand_manifest")


def validate_limits(profile, limits):
    rb.canonical(limits)
    check(profile in PROFILES and type(limits) is dict)
    keys = COMMON_KEYS | ({"max_item_amount_atomic"} if profile == ATOMIC else set())
    check(set(limits) == keys)
    check(limits["network"] == NETWORK and limits["asset"] == ASSET)
    for key in keys:
        if key.startswith("max_"):
            _uint(limits[key])
    _address(limits["recipient"])
    _address(limits["fee_payer"])
    check(limits["recipient"] != limits["fee_payer"])
    hashes = limits["job_hashes"]
    check(type(hashes) is list and 2 <= len(hashes) <= (15 if profile == ATOMIC else 64))
    check(all(type(h) is str and re.fullmatch(r"[0-9a-f]{64}", h) for h in hashes))
    return len(hashes)


def validate_quote(quote, payment_count):
    check(type(quote) is dict and set(quote) == {"network", "genesisHash", "genesisId", "transactionCount", "firstValid", "lastValid", "minFeeMicroAlgo", "feePerByteMicroAlgo", "sponsorFeeMicroAlgo", "observedAt", "expiresAt"})
    check(quote["network"] == NETWORK and quote["genesisHash"] == GENESIS and quote["genesisId"] == "mainnet-v1.0")
    check(type(quote["transactionCount"]) is int and quote["transactionCount"] == payment_count + 1)
    first, last = _uint(quote["firstValid"]), _uint(quote["lastValid"])
    check(first <= last <= first + 1000)
    minimum = _uint(quote["minFeeMicroAlgo"])
    check(1000 <= minimum <= 5000 and quote["feePerByteMicroAlgo"] == "0")
    check(_uint(quote["sponsorFeeMicroAlgo"]) == minimum * (payment_count + 1))
    check(type(quote["observedAt"]) is int and quote["observedAt"] > 0 and type(quote["expiresAt"]) is int and quote["observedAt"] < quote["expiresAt"] <= quote["observedAt"] + 60)
    return quote["expiresAt"]


def validate(envelope, context, limits, profile):
    try:
        rb.canonical(envelope)
        count = validate_limits(profile, limits)
        check(context == rb.request_context(context["url"], "GET"))
        check(type(envelope) is dict and set(envelope) <= {"x402Version", "resource", "accepts", "extensions", "error"} and type(envelope.get("x402Version")) is int and envelope["x402Version"] == 2)
        resource = envelope.get("resource")
        check(type(resource) is dict and set(resource) <= {"url", "mimeType", "description"} and resource.get("url") == context["url"] and all(type(v) is str and len(v.encode()) <= 4096 for v in resource.values()))
        check("error" not in envelope or type(envelope["error"]) is str and len(envelope["error"].encode()) <= 1024)
        check(type(envelope.get("accepts")) is list and len(envelope["accepts"]) == 1)
        req = envelope["accepts"][0]
        check(type(req) is dict and set(req) == {"scheme", "network", "asset", "amount", "payTo", "maxTimeoutSeconds", "extra"})
        check(req["scheme"] == "exact" and req["network"] == NETWORK and req["asset"] == ASSET and req["payTo"] == limits["recipient"])
        check(type(req["maxTimeoutSeconds"]) is int and 1 <= req["maxTimeoutSeconds"] <= 300)
        extra = req["extra"]
        check(type(extra) is dict and set(extra) <= {"feePayer", "decimals"} and extra.get("feePayer") == limits["fee_payer"] and ("decimals" not in extra or type(extra["decimals"]) is int and extra["decimals"] == 6))
        amount = _uint(req["amount"])
        payments = count if profile == ATOMIC else 1
        total = str(amount * payments)
        _uint(total)
        check(type(envelope.get("extensions")) is dict and set(envelope["extensions"]) == {EXTENSION})
        manifest = envelope["extensions"][EXTENSION]
        check(type(manifest) is dict and "feeQuote" in manifest)
        quote = manifest["feeQuote"]
        validate_quote(quote, payments)
        expected = {"version": 2, "profile": profile, "network": NETWORK, "asset": ASSET, "recipient": req["payTo"], "resource": context["url"], "requestHash": rb.digest(context), "jobCount": count, "paymentCount": payments, "paymentAmount": req["amount"], "perJobAmount": req["amount"] if profile == ATOMIC else None, "totalAmount": total, "paymentIndices": list(range(1, payments + 1)), "sponsorIndex": 0, "feePayer": extra["feePayer"], "jobHashes": limits["job_hashes"], "feeQuote": quote}
        check(rb.canonical(manifest) == rb.canonical(expected))
        check(int(total) <= _uint(limits["max_total_amount_atomic"]) and _uint(quote["sponsorFeeMicroAlgo"]) <= _uint(limits["max_sponsor_fee_micro_algo"]))
        if profile == ATOMIC:
            check(amount <= _uint(limits["max_item_amount_atomic"]))
        return expected
    except (ValueError, KeyError, TypeError, UnicodeError, OverflowError):
        raise rb.BindingError("unsupported_algorand_manifest") from None


def validate_atomic(envelope, context, limits):
    return validate(envelope, context, limits, ATOMIC)


def validate_invoice(envelope, context, limits):
    return validate(envelope, context, limits, INVOICE)

File: sdk/route-guard/batch-profiles/algorand-manifest.mjs
// Version2 payment manifests. No network, wallet access or implicit per-job pricing.
import { createHash } from "node:crypto";
const uint = (value) => {
  if (
    typeof value !== "string" ||
    !/^[1-9][0-9]{0,19}$/.test(value) ||
    BigInt(value) > 2n ** 64n - 1n
  )
    throw new Error("unsupported_algorand_manifest");
  return BigInt(value);
};
const NETWORK = "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=",
  ASSET = "31566704",
  EXTENSION = "402signal-atomic-batch";
const check = (x) => {
  if (!x) throw new Error("unsupported_algorand_manifest");
};
const canonical = (x) => {
  if (x === null || typeof x === "boolean" || typeof x === "string")
    return JSON.stringify(x);
  if (Number.isSafeInteger(x)) return String(x);
  if (Array.isArray(x)) return "[" + x.map(canonical).join(",") + "]";
  check(x && Object.getPrototypeOf(x) === Object.prototype);
  return (
    "{" +
    Object.keys(x)
      .sort()
      .map((k) => JSON.stringify(k) + ":" + canonical(x[k]))
      .join(",") +
    "}"
  );
};
const keys = (o, w) =>
  o &&
  Object.getPrototypeOf(o) === Object.prototype &&
  Object.keys(o).sort().join(",") === [...w].sort().join(",");
const address = (s) => {
  check(typeof s === "string" && /^[A-Z2-7]{58}$/.test(s));
  let bits = 0,
    value = 0;
  const out = [];
  for (const c of s) {
    value = (value << 5) | "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567".indexOf(c);
    bits += 5;
    if (bits >= 8) {
      bits -= 8;
      out.push((value >>> bits) & 255);
    }
  }
  const b = Buffer.from(out);
  check(
    (value & ((1 << bits) - 1)) === 0 &&
      createHash("sha512-256")
        .update(b.subarray(0, 32))
        .digest()
        .subarray(-4)
        .equals(b.subarray(32)),
  );
};
export const ATOMIC = "algorand-atomic-multi-item-v1";
export const INVOICE = "algorand-aggregate-invoice-v1";
export function validateAlgorandManifestLimits(profile, l) {
  check(profile === ATOMIC || profile === INVOICE);
  check(
    keys(l, [
      "network",
      "asset",
      "recipient",
      "fee_payer",
      "max_total_amount_atomic",
      "max_sponsor_fee_micro_algo",
      "job_hashes",
      ...(profile === ATOMIC ? ["max_item_amount_atomic"] : []),
    ]),
  );
  check(l.network === NETWORK && l.asset === ASSET);
  for (const [k, v] of Object.entries(l)) if (k.startsWith("max_")) uint(v);
  address(l.recipient);
  address(l.fee_payer);
  check(l.recipient !== l.fee_payer);
  check(
    Array.isArray(l.job_hashes) &&
      l.job_hashes.length >= 2 &&
      l.job_hashes.length <= (profile === ATOMIC ? 15 : 64) &&
      l.job_hashes.every(
        (h) => typeof h === "string" && /^[0-9a-f]{64}$/.test(h),
      ),
  );
  return l.job_hashes.length;
}
export function validateAlgorandFeeQuote(q, payments) {
  check(
    keys(q, [
      "network",
      "genesisHash",
      "genesisId",
      "transactionCount",
      "firstValid",
      "lastValid",
      "minFeeMicroAlgo",
      "feePerByteMicroAlgo",
      "sponsorFeeMicroAlgo",
      "observedAt",
      "expiresAt",
    ]),
  );
  check(
    q.network === NETWORK &&
      q.genesisHash === "wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=" &&
      q.genesisId === "mainnet-v1.0" &&
      q.transactionCount === payments + 1,
  );
  const first = uint(q.firstValid),
    last = uint(q.lastValid),
    minimum = uint(q.minFeeMicroAlgo);
  check(
    first <= last &&
      last <= first + 1000n &&
      minimum >= 1000n &&
      minimum <= 5000n &&
      q.feePerByteMicroAlgo === "0",
  );
  check(uint(q.sponsorFeeMicroAlgo) === minimum * BigInt(payments + 1));
  check(
    Number.isSafeInteger(q.observedAt) &&
      q.observedAt > 0 &&
      Number.isSafeInteger(q.expiresAt) &&
      q.expiresAt > q.observedAt &&
      q.expiresAt <= q.observedAt + 60,
  );
  return q.expiresAt;
}
export function validateAlgorandManifestProfile(e, ctx, l, profile) {
  try {
    canonical(e);
    canonical(l);
    const count = validateAlgorandManifestLimits(profile, l);
    check(
      keys(ctx, ["url", "method", "body_sha256"]) &&
        ctx.method === "GET" &&
        ctx.body_sha256 === createHash("sha256").update("").digest("hex"),
    );
    check(
      typeof ctx.url === "string" &&
        ctx.url.length <= 4096 &&
        /^[\x21-\x7e]+$/.test(ctx.url) &&
        !/[\\#]/.test(ctx.url),
    );
    const url = new URL(ctx.url);
    check(
      ctx.url.startsWith("https://") &&
        url.protocol === "https:" &&
        url.hostname &&
        !url.username &&
        !url.password &&
        !url.hash &&
        !url.port,
    );
    check(
      e &&
        Object.keys(e).every((k) =>
          [
            "x402Version",
            "resource",
            "accepts",
            "extensions",
            "error",
          ].includes(k),
        ) &&
        e.x402Version === 2,
    );
    check(
      e.resource &&
        Object.keys(e.resource).every((k) =>
          ["url", "mimeType", "description"].includes(k),
        ) &&
        e.resource.url === ctx.url &&
        Object.values(e.resource).every(
          (v) => typeof v === "string" && Buffer.byteLength(v, "utf8") <= 4096,
        ),
    );
    check(
      !("error" in e) ||
        (typeof e.error === "string" &&
          Buffer.byteLength(e.error, "utf8") <= 1024),
    );
    check(Array.isArray(e.accepts) && e.accepts.length === 1);
    const req = e.accepts[0];
    check(
      keys(req, [
        "scheme",
        "network",
        "asset",
        "amount",
        "payTo",
        "maxTimeoutSeconds",
        "extra",
      ]) &&
        req.scheme === "exact" &&
        req.network === NETWORK &&
        req.asset === ASSET &&
        req.payTo === l.recipient &&
        Number.isSafeInteger(req.maxTimeoutSeconds) &&
        req.maxTimeoutSeconds >= 1 &&
        req.maxTimeoutSeconds <= 300,
    );
    const extra = req.extra;
    check(
      extra &&
        Object.keys(extra).every((k) => ["feePayer", "decimals"].includes(k)) &&
        extra.feePayer === l.fee_payer &&
        (!("decimals" in extra) || extra.decimals === 6),
    );
    const amount = uint(req.amount),
      payments = profile === ATOMIC ? count : 1,
      total = String(amount * BigInt(payments));
    uint(total);
    check(keys(e.extensions, [EXTENSION]));
    const manifest = e.extensions[EXTENSION];
    check(manifest && Object.hasOwn(manifest, "feeQuote"));
    const quote = manifest.feeQuote;
    validateAlgorandFeeQuote(quote, payments);
    const expected = {
      version: 2,
      profile,
      network: NETWORK,
      asset: ASSET,
      recipient: req.payTo,
      resource: ctx.url,
      requestHash: createHash("sha256").update(canonical(ctx)).digest("hex"),
      jobCount: count,
      paymentCount: payments,
      paymentAmount: req.amount,
      perJobAmount: profile === ATOMIC ? req.amount : null,
      totalAmount: total,
      paymentIndices: Array.from({ length: payments }, (_, i) => i + 1),
      sponsorIndex: 0,
      feePayer: extra.feePayer,
      jobHashes: l.job_hashes,
      feeQuote: quote,
    };
    check(canonical(manifest) === canonical(expected));
    check(
      BigInt(total) <= uint(l.max_total_amount_atomic) &&
        uint(quote.sponsorFeeMicroAlgo) <= uint(l.max_sponsor_fee_micro_algo),
    );
    if (profile === ATOMIC) check(amount <= uint(l.max_item_amount_atomic));
    return expected;
  } catch {
    throw new Error("unsupported_algorand_manifest");
  }
}
export const validateAlgorandAtomicMultiProfile = (e, c, l) =>
  validateAlgorandManifestProfile(e, c, l, ATOMIC);
export const validateAlgorandInvoiceProfile = (e, c, l) =>
  validateAlgorandManifestProfile(e, c, l, INVOICE);
